etor/src/main/java/gov/hhs/cdc/trustedintermediary/external/reportstream/ReportStreamEndpointClient.java

```diff
@@ -0,0 +1,195 @@
+                            300,
+                            ourPrivateKey);
+            String body = composeAuthRequestBody(senderToken);
+            response = client.post(RS_AUTH_API_URL, RS_AUTH_API_HEADERS, body);
+        } catch (Exception e) {
+            throw new ReportStreamEndpointClientException(
+                    "Error getting the API token from ReportStream", e);
+        }
+        // only cache our private key if we successfully authenticate to RS
+        cacheOurPrivateKeyIfNotCachedAlready(ourPrivateKey);
+
+        return response;
+    }
+
+    public String requestWatersEndpoint(@Nonnull String body, @Nonnull String bearerToken)
+            throws ReportStreamEndpointClientException {
+        logger.logInfo("Sending payload to ReportStream");
+
+        Map<String, String> headers =
+                Map.of(
+                        "Authorization",
+                        "Bearer " + bearerToken,
+                        "client",
+                        CLIENT_NAME,
+                        "Content-Type",
+                        "application/fhir+ndjson");
+
+        try {
+            return client.post(RS_WATERS_API_URL, headers, body);
+        } catch (HttpClientException e) {
+            throw new ReportStreamEndpointClientException(
+                    "Error POSTing the payload to ReportStream", e);
+        }
+    }
+
+    public String requestHistoryEndpoint(@Nonnull String submissionId, @Nonnull String bearerToken)
+            throws ReportStreamEndpointClientException {
+        logger.logInfo("Requesting history API from ReportStream");
+
+        Map<String, String> headers = Map.of("Authorization", "Bearer " + bearerToken);
+
+        try {
+            var url = RS_HISTORY_API_URL.replace("{id}", submissionId);
+            return client.get(url, headers);
+        } catch (HttpClientException e) {
+            throw new ReportStreamEndpointClientException(
+                    "Error GETing the history from ReportStream", e);
+        }
+    }
+
+    protected String requestToken() throws ReportStreamEndpointClientException {
+        logger.logInfo("Requesting token from ReportStream");
+
+        String response = requestAuthEndpoint();
+        try {
+            Map<String, String> responseObject =
+                    formatter.convertJsonToObject(response, new TypeReference<>() {});
+            return responseObject.get("access_token");
+        } catch (FormatterProcessingException e) {
+            throw new ReportStreamEndpointClientException(
+                    "Unable to extract access_token from response", e);
+        }
+    }
+
+    public String getRsToken() throws ReportStreamEndpointClientException {
+        logger.logInfo("Looking up ReportStream token");
+
+        var token = cache.get(RS_TOKEN_CACHE_ID);
+
+        if (token != null && isValidToken(token)) {
+            logger.logDebug("valid cache token");
+            return token;
+        }
```

Comment:
Would you need an else condition here to remove an invalid token? Otherwise in theory this map will be growing forever with expired tokens

Reply:
Sounds good!